src/Common/UpgradeNotification/UpgradeNotificationHelper.cs

```diff
@@ -0,0 +1,259 @@
+﻿using Microsoft.Azure.PowerShell.Common.Config;
+using Microsoft.WindowsAzure.Commands.Common;
+using Newtonsoft.Json;
+using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.IO;
+using System.Management.Automation;
+using System.Threading;
+
+namespace Microsoft.Azure.PowerShell.Common.Share.UpgradeNotification
+{
+    public class UpgradeNotificationHelper
+    {
+        private const string AZPSMigrationGuideLink = "https://go.microsoft.com/fwlink/?linkid=2241373";
+        private const string FrequencyKeyForUpgradeNotification = "VersionUpgradeNotification";
+        private static TimeSpan FrequencyTimeSpanForUpgradeNotification = TimeSpan.FromDays(30);
+
+        private const string FrequencyKeyForUpgradeCheck = "VersionUpgradeCheck";
+        private static TimeSpan FrequencyTimeSpanForUpgradeCheck = TimeSpan.FromDays(2);
+        //temp record file for az module versions
+        private static string AzVersionCacheFile = Path.Combine(
+            Environment.GetFolderPath(Environment.SpecialFolder.UserProfile),
+            ".Azure", "AzModuleVerions.json");
+        private bool hasNotified { get; set; }
+        private Dictionary<string, string> versionDict = null;
+
+        private static UpgradeNotificationHelper _instance;
+
+        private UpgradeNotificationHelper()
+        {
+            try
+            {
+                // load temp record file to versionDict
+                if (File.Exists(AzVersionCacheFile))
+                {
+                    using (StreamReader sr = new StreamReader(new FileStream(AzVersionCacheFile, FileMode.Open, FileAccess.Read, FileShare.None)))
+                    {
+                        versionDict = JsonConvert.DeserializeObject<Dictionary<string, string>>(sr.ReadToEnd());
+                    }
+                }
+            }
+            catch (Exception)
+            {
+                versionDict = null;
+            }
+        }
+
+        public static UpgradeNotificationHelper GetInstance()
+        {
+            if (_instance == null)
+            {
+                _instance = new UpgradeNotificationHelper();
+            }
+            return _instance;
+        }
+
+        public void WriteWarningMessageForVersionUpgrade(Microsoft.WindowsAzure.Commands.Utilities.Common.AzurePSCmdlet cmdlet, AzurePSQoSEvent _qosEvent, IConfigManager configManager, IFrequencyService frequencyService) {
+            _qosEvent.HigherVersionsChecked = false;
+            _qosEvent.UpgradeNotificationPrompted = false;
+
+            try
+            {
+                //disabled by az config, skip
+                if (configManager!=null&& configManager.GetConfigValue<bool>(ConfigKeysForCommon.CheckForUpgrade).Equals(false))
```

Comment:
Currently, it appears that customers can only disable notifications before importing the Az module. Disabling notifications after setting the configuration could be a more user-friendly approach.

Reply:
If customers want to opt-out the feature before importing Azure PowerShell modules, they can set the env property "AZUREPS_CHECK_FOR_UPGRADE".